I-ALiRT - updated format of dynamodb

laspsandoval opened the pull request:
# Change Summary

## Overview
This PR is based on feedback from Eric Christian. He said that he thinks that people will just want to query all of the latest data products regardless of the data product name. Therefore, I am the partition key = apid, kept the same sort key (met), and first global secondary index (insert_time), and then added a new global secondary index (data_product_name). 

## Updated Files
- ialirt_ingest_lambda_construct.py
   - Made partition key = apid
   - Added a global secondary index = data_product_name
- ialirt_ingest.py
   - Updated the lambda with the new format

## Testing
- test_ialirt_ingest_lambda_construct.py
- test_ialirt_ingest.py
- conftest.py

laspsandoval committed:
updated format of dynamodb

laspsandoval committed:
update docstring

maxinelasp commented:
Will all the users of this know all the APIDs? I'm worried that this will require other services to maintain an up-to-date list of APID-> product name mappings. I'm also not sure I understand how switching to APID addresses the comment about accessing all the recent products at once... wouldn't that mean the GSI is sufficent?

maxinelasp commented:
Or, maybe I'm showing my ignorance here, can the user query based only off the GSI?

torimarbois commented:
What does APID mean?

laspsandoval commented:
> Will all the users of this know all the APIDs? I'm worried that this will require other services to maintain an up-to-date list of APID-> product name mappings. I'm also not sure I understand how switching to APID addresses the comment about accessing all the recent products at once... wouldn't that mean the GSI is sufficent?

There is only a single I-ALiRT APID. The thing is that the partition key needs to be constant. So I could change it to "IMAP" or something. It really doesn't matter what it is, it just can't change. Because if it does change then you won't be able to query all of the data within a given timerange. I used the APID since that was what we did for the L0 database.

laspsandoval commented:
> What does APID mean?

Application Process Identifier (APID). It's what is assigned to each type of packet. For I-ALiRT it is equal to 478.

laspsandoval commented:
> Or, maybe I'm showing my ignorance here, can the user query based only off the GSI?

Excellent question! In the beginning I had to wrap my head around this too.  You always have to use the partition key (apid). And then you can use the sort key or the global secondary indices to filter the data.

Here is are some examples:

    response = ingest_table.query(
        KeyConditionExpression=Key("apid").eq(478) & Key("met").between(100, 123)
    )

    response = ingest_table.query(
        IndexName="ingest_time",
        KeyConditionExpression=Key("apid").eq(478)
        & Key("ingest_time").begins_with("2021-01"),
    )

    response = algorithm_table.query(
        KeyConditionExpression=Key("apid").eq(478) & Key("met").between(100, 123),
        FilterExpression=Attr("product_name").eq("hit_product_1"),
    )

maxinelasp reviewed (approved):
Thanks for explaining! This looks good to me.
